packages/test/snapshots/src/snapshotStorageService.ts

```diff
@@ -0,0 +1,108 @@
+/*!
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { assert } from "@fluidframework/common-utils";
+import { IDocumentStorageService, ISummaryContext } from "@fluidframework/driver-definitions";
+import { buildSnapshotTree } from "@fluidframework/driver-utils";
+import {
+    ICreateBlobResponse,
+    ISummaryHandle,
+    ISummaryTree,
+    ITree,
+    IVersion,
+    TreeEntry,
+} from "@fluidframework/protocol-definitions";
+import {
+    FileSnapshotReader,
+    IFileSnapshot,
+} from "@fluidframework/replay-driver";
+
+/**
+ * This storage service provides the following functionalities:
+ * - It can be used to load a container with a saved snapshot in `IFileSnapshot` format.
+ * - When a snapshot is written, it calls a callback and provide the snapshot to it in `IFileSnapshot` format. The
+ *   callback should be provided when creating the storage service.
+ */
+export class SnapshotStorageService extends FileSnapshotReader implements IDocumentStorageService {
+    public constructor(
+        json: IFileSnapshot,
+        private readonly snapshotCb: (snapshot: IFileSnapshot) => void,
+    ) {
+        super(json);
+    }
+
+    public async uploadSummaryWithContext(summary: ISummaryTree, context: ISummaryContext): Promise<string> {
+        return Promise.reject(new Error("Invalid operation"));
+    }
+
+    public async createBlob(file: ArrayBufferLike): Promise<ICreateBlobResponse> {
+        return Promise.reject(new Error("Invalid operation"));
+    }
+
+    public async readBlob(blobId) {
+        return Promise.reject(new Error("Invalid operation"));
+    }
+
+    public async downloadSummary(handle: ISummaryHandle): Promise<ISummaryTree> {
+        return Promise.reject(new Error("Invalid operation"));
+    }
+
+    public get repositoryUrl(): string {
+        throw new Error("Invalid operation");
+    }
+
+    public async write(
+        tree: ITree,
+        parents: string[],
+        message: string,
+        ref: string,
+    ): Promise<IVersion> {
+        let dataStoreName = ref ? ref : "container";
+        if (tree && tree.entries) {
+            tree.entries.forEach((entry) => {
+                if (entry.path === ".component" && entry.type === TreeEntry.Blob) {
+                    const content = entry.value.contents.split(":");
+                    if (content[0] === `{"pkg"`) {
+                        dataStoreName = content[1].substring(1, content[1].lastIndexOf(`"`));
+                    }
+                }
+            });
+        }
+
+        // Remove null ids from the tree before calling the callback to notify the new snapshot. This is requried
+        // because the saved reference snapshots have the null ids removed.
+        removeNullTreIds(tree);
```

Comment:
```suggestion
        removeNullTreeIds(tree);
```

Reply:
Looks like this needs to be fixed here as well - https://github.com/microsoft/FluidFramework/blob/main/packages/drivers/file-driver/src/fileDocumentStorageService.ts